sktime/clustering/partitioning/_averaging_metrics.py

```diff
@@ -0,0 +1,220 @@
+     * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+     * GNU General Public License for more details.
+     *
+     * You should have received a copy of the GNU General Public License
+     * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+     ******************************************************************************/
+    TODO: This is basically copied from
+    https://github.com/fpetitjean/DBA/blob/master/DBA.py
+    This need massive optimisations and likely needs to be done in
+    cython. This is basically doing its own implementation of dtw
+    because it needs the cost matrix from the dtw calculation
+    and the current sktime one doesnt return that so need to figure
+    that out
+    """
+
+    def __init__(self, series: NumpyArray, n_iterations: int = 10):
+        super(BarycenterAveraging, self).__init__(series, n_iterations)
+
+    def average(self):
+        """
+        Method used to get the barycenter average
+
+        Returns
+        -------
+        Numpy_Array:
+            Array containing the estimated average
+        """
+        max_length = reduce(max, map(len, self.series))
+
+        cost_mat = np.zeros((max_length, max_length))
+        delta_mat = np.zeros((max_length, max_length))
+        path_mat = np.zeros((max_length, max_length), dtype=np.int8)
+
+        medoid_ind = self._approximate_medoid_index(cost_mat, delta_mat)
+        center = self.series[medoid_ind]
+
+        for _ in range(self.n_iterations):
+            center = self._DBA_update(center, cost_mat, path_mat, delta_mat)
+
+        return center
+
+    def _approximate_medoid_index(self, cost_mat, delta_mat):
+        if len(self.series) <= 50:
+            indices = range(0, len(self.series))
+        else:
+            indices = np.random.choice(range(0, len(self.series)), 50, replace=False)
+
+        medoid_ind = -1
+        best_ss = 1e20
+        for index_candidate in indices:
+            candidate = self.series[index_candidate]
+            ss = self._sum_of_squares(candidate, cost_mat, delta_mat)
+            if medoid_ind == -1 or ss < best_ss:
+                best_ss = ss
+                medoid_ind = index_candidate
+        return medoid_ind
+
+    def _sum_of_squares(self, s, cost_mat, delta_mat):
+        return sum(
+            map(lambda t: self._squared_DTW(s, t, cost_mat, delta_mat), self.series)
+        )
+
+    def _DTW(self, s, t, cost_mat, delta_mat):
+        return np.sqrt(self._squared_DTW(s, t, cost_mat, delta_mat))
+
+    def _squared_DTW(self, s, t, cost_mat, delta_mat):
+        s_len = len(s)
+        t_len = len(t)
+        self._fill_delta_mat_dtw(s, t, delta_mat)
+        cost_mat[0, 0] = delta_mat[0, 0]
+        for i in range(1, s_len):
+            cost_mat[i, 0] = cost_mat[i - 1, 0] + delta_mat[i, 0]
+
+        for j in range(1, t_len):
+            cost_mat[0, j] = cost_mat[0, j - 1] + delta_mat[0, j]
+
+        for i in range(1, s_len):
```

Comment:
For later improvements, we could either try to vectorise these loops or use numba.